internal/scheduler/metrics/metrics.go

```diff
@@ -392,3 +461,64 @@ func (m *Metrics) updateCounterVecFromJob(vec *prometheus.CounterVec, labels []s
 
 	return nil
 }
+
+// updateResourceSecondsCounterVec is a helper method to increment vector counters by the number of seconds per resource a jobs has consumed in a given state.
+func (m *Metrics) updateResourceSecondsCounterVec(vec *prometheus.CounterVec, labels []string, job *jobdb.Job, stateTime, priorStateTime *time.Time) error {
+	i := len(labels)
+
+	// Number of jobs.
+	labels = append(labels, jobsResourceLabel)
+	stateDuration := stateDuration(stateTime, priorStateTime).Seconds()
+	if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+		return err
+	} else {
+		c.Add(stateDuration)
+	}
+	requests := job.GetResourceRequirements().Requests
+	for _, resourceName := range m.config.TrackedResourceNames {
+		labels[i] = string(resourceName)
+		if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+			return err
+		} else {
+			q := requests[resourceName]
+			v := float64(q.MilliValue()) / 1000
+			c.Add(v * stateDuration)
+		}
+	}
+	return nil
+}
+
+func stateDuration(start *time.Time, end *time.Time) time.Duration {
+	if start != nil && end != nil {
+		return end.Sub(*start)
+	}
+	return time.Duration(0)
+}
+
+func getPriorState(run *jobdb.JobRun, stateTime *time.Time) (prior string, priorTime *time.Time) {
+	if stateTime == nil {
+		return "", nil
+	}
+	stateToTime := map[string]*time.Time{
```

Comment:
Shouldn't need to create a map here. I'm concerned it will cause unnecessary GC pressure, which we've had issues with in the past. Just check all of these one by one and return the state/time of the largest timestamp smaller than stateTime?